### ISSUE TITLE
deriveBifunctor fails on Compose

### ISSUE THREAD

**Issue by ocharles** [ekmett/bifunctors#76]
I can currently write the following:

```haskell
data Hello x a = Hello { foo :: Maybe [a] }
deriveBifunctor ''Hello
```

But I cannot write the following equivalent definition:

```haskell
data Hello x a = Hello { foo :: Compose Maybe [] a }
deriveBifunctor ''Helo
```

It is rejected with:

```
src/Language/Haskell/GHC/Token.hs:18:1: error:
    • Couldn't match kind ‘* -> *’ with ‘*’
      When matching types
        p0 :: * -> * -> *
        Compose Maybe :: (* -> *) -> * -> *
      Expected type: p0 b0 c
        Actual type: Compose Maybe [] c
    • In the third argument of ‘Data.Bifunctor.bimap’, namely
        ‘_arg1_aT2x’
      In the first argument of ‘Hello’, namely
        ‘(((Data.Bifunctor.bimap (\ x_aT2D -> x_aT2D)) g_aT2o) _arg1_aT2x)’
      In the expression:
        Hello
          (((Data.Bifunctor.bimap (\ x_aT2D -> x_aT2D)) g_aT2o) _arg1_aT2x)
```

Could this be made to work?

**Comment by RyanGlScott** [comment]
This is an interesting problem. The reason that the second version of `Hello` fails at present is due to the rather simple-minded way that `deriveBifunctor` generates code when it detects an application of a type constructor in a field:

1. If a type constructor has one parameter (e.g., `Maybe a`), then `fmap` it under the assumption that the type constructor has a `Functor` instance.
2. If a type constructor has two or more parameters (e.g., `Either a b`), then `bimap` it under the assumption that the type constructor has a `Bifunctor` instance.

This works for a large swath of use cases, but it's clearly not perfect. Obviously, not every type constructor with two or more parameters has a `Bifunctor` instance. To make things even more annoying, there are some type constructors that with two or more parameters that don't have a `Bifunctor` instance but _do_ have a `Functor` instance, such as `Compose`! Moreover, your first `Hello` example would Just Work™ if we had generated the code `fmap g arg1` instead of `bimap id g arg1`.

I can see two potential ways to resolve this problem:

1. We can revise `deriveBifunctor` so that if we encounter a type constructor with two or more type parameters that only mentions one of the data type's last two type variables in the _very last_ parameter, then use `fmap` instead of `bimap`. For instance, if we encounter `Compose Maybe [] a`, then because `[]` does not mention `x` or `a`, we simply call `fmap`. This would allow the second version of `Hello` to work.

   This approach does come with a downside, however: it would require breaking changes in certain situations. Consider the following example:

   ```hs
   newtype T p a b = MkT (p Int b)
   deriveBifunctor ''T
   ```

   With today's `deriveBifunctor`, the instance that it would generate is:

   ```hs
   instance Bifunctor p => Bifunctor (T p) where ...
   ```

   With the revised version of `deriveBifunctor`, however, this would instead be generated:

   ```hs
   instance Functor (p Int) => Bifunctor (T p) where ...
   ```

   As a consequence, any existing uses of the `Bifunctor` instance for `T` will break unless `p Int` happens to be an instance of `Functor`. (When prototyping this idea, I had several examples in the `bifunctors` test suite break due to this issue, so it's not just a hypothetical concern.) Moreover, this code would also break unless `UndecidableInstances` were enabled.
2. Declare that generating the right code for the second version of `Hello` is out-of-scope for `deriveBifunctor`, and instead recommend people to use something like this idiom:

   ```hs
   {-# LANGUAGE DerivingVia #-}
   import Data.Bifunctor.Joker

   newtype Hello x a = Hello { foo :: Compose Maybe [] a }
     deriving Bifunctor via Joker (Compose Maybe [])
   ```

**Comment by ocharles** [comment]
Thanks for the detailed response. I'd love for 2. to be an option, but unfortunately my type has more than one field. The actual type in question is:

```haskell
data Token token ann =
  Token
    { priorComments :: Compose [] ( Annotated Comment ) ann
    , token :: Annotated token ann
    }
```

Which I don't think can be used with DerivingVia.

I think given the complexity this might just be a won't fix. After all, the manual instance is:

```haskell
instance Bifunctor Token where
  bimap f g ( Token x y ) =
    Token ( fmap g x ) ( bimap f g y )
```

**Comment by RyanGlScott** [comment]
> I'd love for 2. to be an option, but unfortunately my type has more than one field. The actual type in question is:
> 
> ```haskell
> data Token token ann =
>   Token
>     { priorComments :: Compose [] ( Annotated Comment ) ann
>     , token :: Annotated token ann
>     }
> ```
> 
> Which I don't think can be used with DerivingVia.

Ah, quite right. As is often the case, half the battle is picking the right newtypes. If you _really_ want to make this work with today's `deriveBifunctor`, you could always swap out `Compose [] (Annotated Comment) ann` with `Tannen [] Annotated Comment ann` (where `Tannen` comes from [`Data.Bifunctor.Tannen`](https://hackage.haskell.org/package/bifunctors-5.5.4/docs/Data-Bifunctor-Tannen.html#t:Tannen)). But as you observe, implementing the `Bifunctor` instance by hand works just as well.

> I think given the complexity this might just be a won't fix.

To be clear, I don't think implementing option (1) would be difficult—it would just be something that would require a major version bump. Whether the payoff would be worth the code churn is unclear to me at present, but if other people chime in saying that they want this feature too, perhaps I could be convinced :)

**Comment by ocharles** [comment]
`Tannen` works... but then I can't derive `Show1` using `deriving-compat` :laughing: :

```
src/Language/Haskell/GHC/Token.hs:36:1: error:
    • Could not deduce (Data.Functor.Classes.Show1
                          (Tannen [] Annotated Comment))
        arising from a use of ‘Data.Functor.Classes.liftShowsPrec’
      from the context: Show token
        bound by the instance declaration
        at src/Language/Haskell/GHC/Token.hs:36:1-19
    • In the first argument of ‘(.)’, namely
        ‘(((Data.Functor.Classes.liftShowsPrec sp1_afLg) sl1_afLh) 0)
           arg1_afLt’
      In the second argument of ‘(.)’, namely
        ‘((((Data.Functor.Classes.liftShowsPrec sp1_afLg) sl1_afLh) 0)
            arg1_afLt
            . (GHC.Show.showCommaSpace
                 . (showString "token = "
                      . ((((Data.Functor.Classes.liftShowsPrec sp1_afLg) sl1_afLh) 0)
                           arg2_afLu
                           . showChar '}'))))’
      In the second argument of ‘(.)’, namely
        ‘(showString "priorComments = "
            . ((((Data.Functor.Classes.liftShowsPrec sp1_afLg) sl1_afLh) 0)
                 arg1_afLt
                 . (GHC.Show.showCommaSpace
                      . (showString "token = "
                           . ((((Data.Functor.Classes.liftShowsPrec sp1_afLg) sl1_afLh) 0)
                                arg2_afLu
                                . showChar '}')))))’
```

**Comment by RyanGlScott** [comment]
Ah, well that's easy enough to explain: no one has bothered to add a `Show1` instance for `Tannen` yet. I've done so in #77.

**Comment by RyanGlScott** [comment]
I've merged #77. @ocharles, does this make `Tannen` suitable for your purposes?

**Comment by ocharles** [comment]
Sorry Ryan, I moved away from a bifunctor approach now. #77 still looks like a net win though.
